colossalai/tensor/_ops/linear.py

```diff
@@ -4,72 +4,50 @@
 from colossalai.nn.layer.utils import divide
 from colossalai.core import global_context as gpc
 from packaging import version
-from colossalai.tensor import ComputePattern, TensorSpec, ComputePattern, ParallelAction, ColoTensor, ShardPattern
+from colossalai.tensor import ComputePattern, TensorSpec, ComputePattern, ParallelAction, ColoTensor, dist_spec
 from colossalai.tensor.graph import GraphOpNode, GraphGlobalEnv
 
 
 def colo_linear_1Drow(input_tensor: ColoTensor, weight: ColoTensor, bias: ColoTensor) -> ColoTensor:
-    parallel_action = weight.shard_spec.get_action_by_compute_pattern(ComputePattern.TP1DRow_Linear)
+    parallel_action = weight.spec.get_action_by_compute_pattern(ComputePattern.TP1DRow)
     # Input:S[1] x Weight:S[0] = Output:P
     # All-Reduce(Output) + bias = res
     # Input:S[1]
-    if input_tensor.is_gathered():
-        # Not splited yet.
-        assert divide(input_tensor.shape[-1], gpc.tensor_parallel_size) == weight.size(-1), \
-        'Invalid shapes in 1Drow forward: input={}, weight={}. Expected last dim of input {}.'.format(
-        input_tensor.shape, weight.size, weight.size(-1) * gpc.tensor_parallel_size)
-        input_per_partition = split_forward_gather_backward(input_tensor.torch_tensor(),
-                                                            parallel_action.parallel_mode,
-                                                            dim=-1)
-    elif input_tensor.shard_pattern == ShardPattern.Col:
-        # Splited by 1Dcol
-        assert input_tensor.shape[-1] == weight.size(-1), \
-        'Invalid shapes in 1Drow forward: input={}, weight={}. Expected last dim of input {}.'.format(
-        input_tensor.shape, weight.size, weight.size(-1))
-        input_per_partition = input_tensor.torch_tensor()
-    else:
-        raise NotImplementedError
+    input_tensor.to_dist_spec(
+        dist_spec.shard(weight.spec.dist_spec.process_group, [-1], [weight.spec.dist_spec.process_group.size()]))
```

Comment:
The nesting naming is confusion....
a file.and sepc.py also a file containing dist_spec? But dist_spec.py is also a file.

Reply:
why not just make a file named api.py?

Reply:
In short, I think the following usage looks better to developers.
weight.spec.get_process_group()

Reply:
I think the first spec is TensorSpec and the second one is DistSpec. Do we really neead TensorSpec abs?

Reply:
Flattern the folder structure now.

Reply:
TensorSpec includes a action list now.

Reply:
If we move the action list to op rather than parameter, then TensorSpec and DistSpec can be merged into one.

Reply:
It's actually introduced by ComputePattern bound to weight.

Reply:
Use the weight.spec.get_process_group() mentioned by @feifeibear or add a TODO for ComputePattern code refining.

Reply:
`spec.get_process_group()` added.